src/core/mac/mac_frame.cpp

```diff
@@ -734,40 +764,78 @@ void Frame::SetPayloadLength(uint16_t aLength)
 
 uint8_t Frame::SkipSecurityHeaderIndex(void) const
 {
-    uint8_t  index = 0;
+    uint8_t  index = kFcfSize + kDsnSize; // Frame Control field and Sequence Number field
     uint16_t fcf;
 
-    // Frame Control
-    index += kFcfSize;
-    // Sequence Number
-    index += kDsnSize;
-
     VerifyOrExit((index + GetFcsSize()) <= GetPsduLength(), index = kInvalidIndex);
 
     fcf = GetFrameControlField();
+    index += SkipAddrField(fcf);
 
-    // Destination PAN + Address
+    // Security Control + Frame Counter + Key Identifier
+    if ((fcf & kFcfSecurityEnabled) != 0)
+    {
+        VerifyOrExit(index < GetPsduLength(), index = kInvalidIndex);
+        uint8_t securityControl = *(GetPsdu() + index);
+
+        index += kSecurityControlSize + kFrameCounterSize;
+
+        switch (securityControl & kKeyIdModeMask)
+        {
+        case kKeyIdMode0:
+            index += kKeySourceSizeMode0;
+            break;
+
+        case kKeyIdMode1:
+            index += kKeySourceSizeMode1 + kKeyIndexSize;
+            break;
+
+        case kKeyIdMode2:
+            index += kKeySourceSizeMode2 + kKeyIndexSize;
+            break;
+
+        case kKeyIdMode3:
+            index += kKeySourceSizeMode3 + kKeyIndexSize;
+            break;
+        }
+    }
+
+exit:
+    return index;
+}
+
+uint8_t Frame::SkipAddrField(const uint16_t &fcf) const
+{
+    uint8_t offset = 0;
+
+    // Destination PAN
+    if (IsDstPanIdPresent(fcf))
+    {
+        offset += sizeof(PanId);
+    }
+
+    // Destination Address
     switch (fcf & kFcfDstAddrMask)
     {
     case kFcfDstAddrNone:
         break;
 
     case kFcfDstAddrShort:
-        index += sizeof(PanId) + sizeof(ShortAddress);
+        offset += sizeof(ShortAddress);
         break;
 
     case kFcfDstAddrExt:
-        index += sizeof(PanId) + sizeof(ExtAddress);
+        offset += sizeof(ExtAddress);
         break;
 
     default:
-        ExitNow(index = kInvalidIndex);
+        break;
```

Comment:
I would suggest we pass `index` as a both [in] and [out] parameter so that we can update `index` correctly. 
Otherwise the original fault handling code is lost. 
```suggestion
        ExitNow(index = kInvalidIndex);
```